thriftbp/circuit_breaker.go

```diff
@@ -0,0 +1,123 @@
+	FailureThreshold float64
+	// Name for this circuit breaker, mostly used as a prefix to disambiguate logs when multiple cb are used.
+	Name string
+	// EmitStatusMetrics sets whether the failure breaker will regularly update a gauge on the breakers state (closed or open/halfopen)
+	EmitStatusMetrics bool
+	// Logger is the logger to be called when edgecontext parsing failed.
+	Logger log.Wrapper
+	// MaxRequestsHalfOpen represents he Maximum amount of requests that will be allowed through while the breaker
+	// is in half-open state. Unset or 0 is treated as 1
+	MaxRequestsHalfOpen uint32
+	// Interval represents the cyclical period of the 'Closed' state.
+	// If 0, internal counts do not get reset while the breaker remains in the Closed state.
+	Interval time.Duration
+	// Timeout is the duration of the 'Open' state. After an 'Open' timeout duration has passed, the breaker enters 'half-open' state.
+	Timeout time.Duration
+}
+
+// NewFailureRatioBreaker creates a new FailureRatioBreaker with the provided configuration. Creates a new goroutine to emit
+// breaker state metrics if EmitStatusMetrics is set to true.
+func NewFailureRatioBreaker(config CircuitBreakerConfig) FailureRatioBreaker {
+
+	failureBreaker := FailureRatioBreaker{
+		name:              config.Name,
+		minRequestsToTrip: config.MinRequestsToTrip,
+		failureThreshold:  config.FailureThreshold,
+		logger:            config.Logger,
+	}
+	settings := gobreaker.Settings{
+		Name:          config.Name,
+		Interval:      config.Interval,
+		Timeout:       config.Timeout,
+		MaxRequests:   config.MaxRequestsHalfOpen,
+		ReadyToTrip:   failureBreaker.ShouldTripCircuitBreaker,
+		OnStateChange: failureBreaker.stateChanged,
+	}
+
+	failureBreaker.goBreaker = gobreaker.NewCircuitBreaker(settings)
+	if config.EmitStatusMetrics {
+		go failureBreaker.runStatsProducer()
+	}
+	return failureBreaker
+}
+
+func (cb FailureRatioBreaker) runStatsProducer() {
+	circuitBreakerGauge := metricsbp.M.RuntimeGauge(cb.name + "-circuit-breaker-closed")
+
+	tick := time.NewTicker(10 * time.Second)
+	defer tick.Stop()
+	for {
+		select {
+		case <-metricsbp.M.Ctx().Done():
+			return
+		case <-tick.C:
+			if cb.goBreaker.State() == gobreaker.StateOpen {
+				circuitBreakerGauge.Set(0)
+			} else {
+				circuitBreakerGauge.Set(1)
+			}
+		}
+	}
+}
+
+// ThriftMiddleware is a thrift.ClientMiddleware that handles circuit breaking.
+func (cb FailureRatioBreaker) ThriftMiddleware(next thrift.TClient) thrift.TClient {
+	return thrift.WrappedTClient{
+		Wrapped: func(ctx context.Context, method string, args, result thrift.TStruct) error {
+			_, err := cb.goBreaker.Execute(func() (interface{}, error) {
+				return nil, next.Call(ctx, method, args, result)
+			})
+			return err
+		},
+	}
+}
+
+// ShouldTripCircuitBreaker checks if the circuit breaker should be tripped, based on the provided breaker counts.
+func (cb FailureRatioBreaker) ShouldTripCircuitBreaker(counts gobreaker.Counts) bool {
```

Comment:
Do we really want this function to be exported? It's not required to be used by go-breaker library, so exporting it only makes sense if we intend it to be used by some other ways, and I'm not sure we have any other usage for it?

Reply:
Changed it all to mocks, wasnt as bad as I thought.